colorizer_data/tests/test_converter.py

```diff
@@ -81,59 +87,238 @@ def test_handles_simple_csv(tmp_path):
+    csv_data = pd.read_csv(StringIO(csv_content))
+    # Delete all the columns that have feature data
+    csv_data = csv_data.drop(
+        ["Continuous Feature", "Discrete Feature", "Categorical Feature"], axis=1
+    )
+    with pytest.raises(Exception):
+        convert_colorizer_data(csv_data, tmp_path)
+
+
+def test_fails_if_no_objects_exist(tmp_path):
+    csv_content = f"{sample_csv_headers}"
+    csv_data = pd.read_csv(StringIO(csv_content))
+
+    with pytest.raises(Exception):
+        convert_colorizer_data(csv_data, tmp_path)
+
+
+def test_handles_missing_centroid_and_outlier_columns(tmp_path):
+    csv_content = f"{sample_csv_headers}\n{sample_csv_data}"
+    csv_data = pd.read_csv(StringIO(csv_content))
+    csv_data = csv_data.drop(["Centroid X", "Centroid Y", "Outlier"], axis=1)
+    convert_colorizer_data(csv_data, tmp_path, output_format=DataFileType.JSON)
+
+    # Outliers and centroids should not be written
+    assert not os.path.exists(tmp_path / "outliers.json")
+    assert not os.path.exists(tmp_path / "centroids.json")
+    # Data should not be in manifest
+    manifest = {}
+    with open(tmp_path / "manifest.json", "r") as f:
+        manifest = json.load(f)
+        assert "outliers" not in manifest
+        assert "centroids" not in manifest
+
+
+def test_throws_error_if_all_values_are_outliers(tmp_path):
+    csv_content = f"{sample_csv_headers}\n{sample_csv_data}"
+    csv_data = pd.read_csv(StringIO(csv_content))
+    csv_data["Outlier"] = 1
+    with pytest.raises(Exception):
+        convert_colorizer_data(csv_data, tmp_path)
+
+
+def test_uses_id_as_track_if_track_is_missing(tmp_path):
+    csv_content = f"{sample_csv_headers}\n{sample_csv_data}"
+    csv_data = pd.read_csv(StringIO(csv_content))
+    csv_data = csv_data.drop(["Track"], axis=1)
+    convert_colorizer_data(csv_data, tmp_path, output_format=DataFileType.JSON)
+
+    manifest = {}
+    with open(tmp_path / "manifest.json", "r") as f:
+        manifest = json.load(f)
+        print(manifest)
+        assert manifest["tracks"] == "tracks.json"
+        # Uses IDs as track IDs
+        validate_data(tmp_path / "tracks.json", [0, 1, 2, 3])
 
 
 """
 TODO: Test additional edge cases
-- Frame generation
-  - override switch works
-  - detects change in number of objects
-  - detects removal of frames
-  - does not regenerate frames if they already exist
-- Handles missing centroid, outliers, or bounds data
-- Handles backdrop images via column
-- Handles backdrop images via dictionary
+- [x] Frame generation
+  - [x] override switch works
+  - [x] detects change in number of objects
+  - [x] detects removal of frames
+  - [x] does not regenerate frames if they already exist
+  - [x] handles missing times data
+- [x] Handles different data column names
+- [x] Handles missing centroid, outliers, or bounds data
+- [x] Keeps bounds data during frame regeneration
+- [ ] Handles backdrop images via column
+- [ ] Handles backdrop images via dictionary
```

Comment:
you can just remove the ones with x's...they aren't TODOs

Reply:
This gets deleted in the next PR!